### ISSUE TITLE
Radar/Spider/Net Charts

### ISSUE THREAD

**Issue by ccoupe** [Shoes3/shoes3#288]
Here's the latest view on radar charts. Just draws the radial axes and the label of max value of that column. It draws the label poorly but we talk about 'better' after the image.
![radar-2](https://cloud.githubusercontent.com/assets/222691/19916735/b6ef86f4-a082-11e6-95c4-ae14578971fc.png)

In gruft and other source code I've looked at,  our `chart_series` is equivalent  to  a `row` and the value inside the rows are columns. Radar and the like, like to draw/compute down columns.  Shoes currently has no API to specify some things for that orientation that users might find useful.  In the example above, the labels shown are the **computed** max (down the columns) but before we can scale and and draw some data points and connected lines, they need to be scaled between min and max. What's the value at the center for each? Compute min? 0? Something else? Can it be specified by the user and if so, how?

**Comment by ccoupe** [comment]
Sadly, we do need add an extra arg to _plot_ creation when creating a radar. The most natural is an array of arrays. Ick!  Do I have good name - of course not!  Let's call  it  "columns_settings:"  until someone suggests a better name and it only applies to certain chart types and it is mandatory for radar (and perhaps some new chart types in the future).  

`plot {...., chart: "radar", column_settings:  [  <column 0 settings>,  <column 1 settings>,  ...]}`
 What's in a 'column setting'? I think it should be another array or a hash - users choice. Hash is more descriptive but it gets wordy so both should be acceptable  `{ label: <string>, min: <number>, max: <number>, format: <printf_format_string> }` .You might wonder about the format: string. If not given then it defaults to "%4.2f" which is not optimal for anyone who cares about their graph.  I could even imagine other options like alpha: 0.5 if you want the damn things dimmer (I would).

seeking opinions @BackOrder , @passenger94 .

**Comment by BackOrder** [comment]
What this new parameters for? It is normally possible to create radar graph with same kind of plot data as any other graphs. For example, see this [tutorial](https://support.office.com/en-us/article/Present-your-data-in-a-radar-chart-16e20279-eed4-43c2-9bf5-29ff9b10601d) for Microsoft Office.

Now it might be a good idea to support optional parameters for plot. I'm just confused why it would be mandatory for radar graph...

**Comment by BackOrder** [comment]
Also as reference another [tutorial](https://www.extendoffice.com/documents/excel/2135-excel-create-radar-chart.html).

**Comment by ccoupe** [comment]
Excel and the like have the labels for rows and columns (because they have to be there, or you have to add them later in the wizard) Shoes does not have a wizard for creating charts.  Suppose the two values in column 0 are [80, 180]. Are those 0 based? Or should that 80 be the  dot in the center?  What if another  radial/column on the same chart  ranges from  20 to -10. Since  Shoes adds **and draws** (chart_series)  row by row and does not know how many rows will be added later. It draws what it has. Excel and friends know the that before drawing anything.

**Comment by BackOrder** [comment]
My understanding is that radar graph is based on its given values, using minimum and maximum values and evenly stepped. Minimum value is the centre. We may offer optional parameters (unit, scale, step, etc.) for customization. Does it make sense to you?

**Comment by ccoupe** [comment]
It make sense in some cases. And lacking for others.  In the [80, 180] case, you only get two dots when it gets around to drawing the lines and dots - center and the outer radius. What to do if  you know that 180 is not the max score on the test and 80 isn't the worst. Maybe 250 is the best (outer ting and 50 is the worst (center) .  How to specify that?

**Comment by BackOrder** [comment]
Would you mind sharing a sample code of radar graph in Gruff or any other related API (slash language)? Maybe we could learn something from it.

**Comment by ccoupe** [comment]
Crux => matter. Excel is an interactive tool for creating charts. Shoes is declarative. So is gruff (clone it and look at the test cases) or pycha and look at the test cases (hard to find). Look at what gnuplot uses for a DSL  (bonus points if if you can find a radar/net/spider example that makes sense). Matlib? Wolfram? Some Adobe Thing?. Shoes can't be all of the above. Use them if Shoes too much trouble.

**Comment by BackOrder** [comment]
Take a look at [Chart.js](http://www.chartjs.org/docs/) and [rgraph](http://www.rgraph.net/docs/radar.html). They  both have a very clean API. Please, notice Chart.js has an options argument for the special case(s) you are referring to. Otherwise, they both proceed as I stated before.

**Comment by ccoupe** [comment]
There's no doubt that my bottoms up approach is problematic.  Originally, all I wanted was the timeseries chart that I could append RT data to and add and delete series. And then it grew from there.  It's not a goal for Shoes to replace dedicated charting applications (or spreadsheets). Just something simple.  Now, it's not so simple. 

I'm considering (just thinking aloud) that the existing Time series chart and RT stuff should be a different widget (and api) from all the other charts or perhaps an alternative api for all the other other chart types. For now, I'll probably bumble ahead just to finish up and learn some things.

**Comment by BackOrder** [comment]
It would ideally be implemented in Ruby/Shoes level for maximum extensibility but this approach has its challenges.

Regardless of the current flaws or original intents, I assure you this is a very desirable feature. Charting feature is surpringly difficult to find in any language.

**Comment by ccoupe** [comment]
>It would ideally be implemented in Ruby/Shoes level for maximum extensibility but this approach has its challenges.

#266 + gruff for a template. Maybe we can talk @passenger94 into implementing the text_shape.He  know more about that place inside Shoe than I. 

>Regardless of the current flaws or original intents, I assure you this is a very desirable feature. Charting feature is surpringly difficult to find in any language. 

I also observed that empty spot in the OSS desktop universe. It's why I'd like to get this feature 'good enough' for fun and exploration without getting too complex or detailed. 
```ruby
  @values1 = [80,        160,          145,     75,          80] # in k$
  @values2 = [180,        90,           95,     90,          90]
  @columns = [ ["internet", 0, 200], 
               ["television", 0, 200],
               {label: "radio", min: 0, max: "mistake", format: "%3.3f"}  ,
               #["radio", 0, 200],
               ["newspaper", 0, 200],
               ["magazine", 0, 200, "%dk"]
             ]
...
       @grf = plot widget_width, widget_height, title: "Advertising", caption: 
          "Budget Spend" , font: "Helvetica", auto_grid: true,
          default: "skip", background: white, chart: "radar", column_settings: @columns
      end
```
Not completely awful. Extensible at the C level (sort of).

**Comment by ccoupe** [comment]
Getting closer.
![radar-3](https://cloud.githubusercontent.com/assets/222691/20035465/29b2658c-a3ab-11e6-8214-e4e49e07cebb.png)

There are many things to improve but basics are not awful.

**Comment by BackOrder** [comment]
Looking good!

**Comment by ccoupe** [comment]
Slightly more appealing.  Incorrect label positions  but improving. 
![radar-4](https://cloud.githubusercontent.com/assets/222691/20087619/744f7444-a537-11e6-97b8-a231601d3c56.png)

**Comment by BackOrder** [comment]
Are measurement lines also implemented? It seems to go really well so far.

Good news is that features you have implemented lately, radar graph included, will allow me to port one of my prototypes on Shoes and implement a fully functional prototype. Thanks for your continuous efforts!

**Comment by ccoupe** [comment]
Many, many and more details need to be added.  I assume measurement lines are the lines drawn round the axis connecting the xaxis labels. In the tiny chart above there is room for an outer ring connecting all 200 end points  and say one 'ring' at 100' since the examples all start with 0 at the center point. 

Remember, all five of the above axis can have different scales so you need value labels on each of those points which can make for a very busy chart.  And of course, someone will  want options to not do that or only draw two when the room for 10. Or maybe they just want little 'tick' marks instead of labels.  Another API challenge! There really is no end to possible enhancements and feature creep to this and many others charts all of which are very difficult to (re)implement incrementally  in C.  (wouldn't be that much easier in Ruby)

If you can provide a screen shot of what you really need (and the test date), that would help.

**Comment by BackOrder** [comment]
[Statistics Feature for Humintell Products](https://fr.scribd.com/document/28961133/Statistics-for-Humintell-Products) is how I use radar chart in general. This is how OpenOffice/LibreOffice generate basic radar charts.

You may notice the scale is in percent from 0% (center of the chart) to 100% (outward). The scale is the same for every item (7 universal emotions in this case). Each coloured line represent percentage obtained at a precise moment in time (pre-test score, practice score, post-test score). In this document, the pre-test never change once taken, practice is updated as practice goes, post-test is created once post-test in taken (and possibly updated if taken again).

In a nutshell, basic features are fully sufficient for most case scenarios. The extra features you are suggestion are fine but as long as the basic ones are fully implemented, I can use your radar graphs.

Let me check if I could get some data for you... stay tunes.

**Comment by BackOrder** [comment]
![image](https://cloud.githubusercontent.com/assets/8844578/20201260/a081b2bc-a784-11e6-8edb-102a06bf8168.png)

**Comment by ccoupe** [comment]
Thanks.  I'll use that for a for another test case (gr7.rb). We  can get pretty close today.  South of your border, an number of people appear to need help handling their emotions after the recent election.

**Comment by ccoupe** [comment]
I do like other peoples tests.  This one demonstrates just how much work remains. 
![radar-5](https://cloud.githubusercontent.com/assets/222691/20203242/ed6533da-a782-11e6-8dc4-3e45f6217f95.png)

There is a Windows [beta that (hopefully](http:/walkabout.mvmanila.com/public/shoes/shoes-3.3.2-gtk3-32.exe)) can draw that  if you get the Tests/plot/gr7.rb from github.

**Comment by BackOrder** [comment]
Indeed, indeed. The data in you recent commit seems to be on point with my table. Unfortunately, the radar graph is incorrect. It seems some data is mixed up internally. Also, notice my example starts with **Anger** on top as it is the first item.

![image](https://cloud.githubusercontent.com/assets/8844578/20203653/0292394c-a797-11e6-8183-6428f2739cce.png)

**Comment by ccoupe** [comment]
There are many bugs on the _to do_ list. That's one of them.

**Comment by ccoupe** [comment]
Also note that Shoes and Your example are the difference between clockwise and anti-clockwise. Either way will convey the same knowledge, right? (another reason I don't love radar/spider)

**Comment by BackOrder** [comment]
The main objection is that every other engine, API and office/drawing suites is drawing radar graph the same way. It means anyone who used it before will complain and likely to shuffle their own data (at great inconvenience) to get same result as others.

Basically, might as well do the same.

**Comment by ccoupe** [comment]
Back in the days when I managed Software Engineers, They would have said the same thing when I  proposed a different way forward. Convention is paramount! They would push back and sometimes  they were correct and sometimes they pushed back, just because it wasn't done their way . Now, I'm just a SE of dubious skills on a collaborative  OSS project and sometimes I'm the self appointed manager of this "new way to write Ruby GUI's that doesn't it do the conventional  way".

**Comment by BackOrder** [comment]
Haha! You know I earn a living doing different way from others? It however has to make sense, has a rationale behind the change.

So what does your current implementation bring that the conventional way doesn't?

**Comment by ccoupe** [comment]
>So what does your current implementation bring that the conventional way doesn't?

We've gone through this before. Real Time updating (timeseries charts),  add and delete series of data from displayed charts (varies by chart type), very large data sets that would take weeks or months to type by hand and mostly, 2800 lines of C code and partially finished documentation and many man hours  that are not amenable to 'just redo it this way'.

Once we get #266 implemented then **anybody else** can write many many  lines of Ruby/Shoes that does it 'their' way for their api and it won't interfere with the minimalistic C plot widget.

**Comment by BackOrder** [comment]
Well, doesn't the radar graph just need rotation? Shouldn't be a big deal..

**Comment by ccoupe** [comment]
Depends on the C code whether it  is a big deal or not.  I'd like to have the outer labels rotate in the same direction  (known bug)  before I care about which direction they rotate. It makes no difference in user comprehension of what is displayed. You don't have to look very hard to find that is there is no convention for your chosen direction. Useless Quibbblles.

**Comment by ccoupe** [comment]
Some success with measurement lines ( I call them rings in the C code) These examples draw 3 rings. Notice how Shoes detects that there may be different scales needed.
![radar-6](https://cloud.githubusercontent.com/assets/222691/20255657/9d762d40-a9fb-11e6-96c2-19d5576f7131.png) ![radar-7](https://cloud.githubusercontent.com/assets/222691/20255757/72f17b00-a9fc-11e6-8d06-c42e5f081bce.png)

Yes outer labels are wrong.

**Comment by BackOrder** [comment]
It's enjoyable to see the progress. Please, notice the lines from center to outwards are progressively lighter. I suspect each section are redrawn on top of previous sections instead of just once.

**Comment by ccoupe** [comment]
Good eye! Thanks. I was using black+alpha because it was too hard to look up the color of gray I wanted.  Another example of why more eyes than mine are good.

**Comment by ccoupe** [comment]
Turns out to be more that an  alpha misuse.  Fixed. I added an option to the  plot creation api , **rings:** which accepts Ruby _true_, _false_, or an _integer_. T/F maps to 0 or 1 as you might expect.  false or 0 means no radials or interior labels. true or 1 (the default) means to use  Shoes' idea of how many measurement rings should be drawn for the size of of the chart.  >= 2 or greater than 2 means that's how many rings you want. A bit odd, but it does make sense.

If you don't like **rings:** argument name, It's easy to change.

**Comment by BackOrder** [comment]
> Ruby _true, false,_ or an _integer_.

It should be possible to support true/false without having them meaning 0 or 1. Then the integer would actually be meaningful. Also, it would allow to actually use 0 or 1 as number of rings. 

> If you don't like **rings:** argument name, It's easy to change.

The correct term is **grid lines**.

**Comment by ccoupe** [comment]
>It should be possible to support true/false without having them meaning 0 or 1. Then the integer would actually be meaningful. Also, it would allow to actually use 0 or 1 as number of rings. 

It checks for false or  true. if neither it checks for integer. internally it's an stored in one C int, no matter what was used in Shoes/Ruby. 

grid_lines is fine with me.

**Comment by BackOrder** [comment]
> It checks for false or true. if neither it checks for integer. internally it's an stored in one C int, no matter what was used in Shoes/Ruby.

I was under the impression it was possible to check for True or False class, as it would in Ruby.

> grid_lines is fine with me.

Sure, let's go with **grid_lines**. It should be alright as most people will never use it, if the default value is reasonable. Ruby traditionally favours underscore method names instead of camel case. Also, **lines** would be shorter, shorter is normally better, but it seems too broad of a meaning to use it here.

**Comment by ccoupe** [comment]
It is possible to check for class names but its more wordy (plus a function call) compared to checking if its Qfalse or Qtrue C constants.

**Comment by BackOrder** [comment]
> It is possible to check for class names but its more wordy (plus a function call) compared to checking if its Qfalse or Qtrue C constants.

It's not really worth all the troubles, isn't it? So your way is the way.

**Comment by ccoupe** [comment]
Getting Closer on the outer labels and same rotation. Obvious problems - not so obvious how to fix.
![radar-8](https://cloud.githubusercontent.com/assets/222691/20336391/e7a86572-ab86-11e6-8308-b5c4575b4f45.png)

**Comment by BackOrder** [comment]
Here is some pointers for a proper label alignment. I drew an _imaginary_ circle and guide lines that would be bigger than the radar graph. The point between a line and the circle is the anchor. **The anchor should be use to center text on.** Hope this helps.

![image](https://cloud.githubusercontent.com/assets/8844578/20373136/22e087b6-ac3e-11e6-9c14-80186f48181f.png)

**Comment by BackOrder** [comment]
The following one use the same principle but the anchor is used as a left/right centering rather than centering, depending on which side of the graph the text is on. It is right centering for those on the left sided quadrants and left centering for those on the right sided quadrants.

![image](https://cloud.githubusercontent.com/assets/8844578/20373467/1e5b95da-ac40-11e6-9ba2-59ae7e245dc5.png)

**Comment by BackOrder** [comment]
Here is another alignment you might be interested in. Hopefully it inspires you.

![image](https://cloud.githubusercontent.com/assets/8844578/20375853/c6525288-ac4f-11e6-9bce-2cb9388b5e69.png)

**Comment by ccoupe** [comment]
Currently, you last example is close to what I'm attempting.  I'm using the quad calculations from the pie chart where things look ok.  One issue is that the first, top most Anger is quad 0, as is Contempt but they connect to the imaginary text box in different locations.  I can special case that.

**Comment by ccoupe** [comment]
This one looks good enough (for Shoes)
![radar-9](https://cloud.githubusercontent.com/assets/222691/20377785/07fb7878-ac4f-11e6-8af1-908c9d2b7953.png)

There are issues exposed with Test/lplot/gr6.rb which needs some quadrant hints for positioning inner labels.

**Comment by ccoupe** [comment]
Tests/plot/gr6 requires multiple inner labels at the because not all things are scaled the the same. This is damn close to as much as I'm willing to do for Shoes 3.3.2 release. 
![radar-10](https://cloud.githubusercontent.com/assets/222691/20380333/a5802fa8-ac5e-11e6-8838-83aece3bf8ee.png)

**Comment by BackOrder** [comment]
Very good! Take note the space between anger and the radar graph vs space between joy/fear and the radar graph seems different on Test/lplot/gr6.rb. Most likely joy/fear are too close. Other than that, the rest seems on point.

**Comment by ccoupe** [comment]
Funny you should mention that. The code is 
```C
// TODO: user specified radius multipler
double radius = chart->radius * 1.15; // extend it to draw outside
```
I think if was set to 1.10 for the gr7 plot.  So, another something: <number> has to be added to the API

**Comment by BackOrder** [comment]
By the way, I also notice labels/data are drawn clockwise. It seems every engine draw counter-clockwise. What do you think?

**Comment by ccoupe** [comment]
>  labels/data are drawn clockwise

We've been here before - dueling conventions. 
Oddly enough,  I use the most common  C idiom of `for (i = 0; i < max; i++)` instead of the error prone `for (i= max; i > 1;  i--)` or is it ` for (i = max-1; i >0;  i--)`?

>What do you think?
I think I'll wait for someone else to cross that mine field and show me their C code.

**Comment by BackOrder** [comment]
`for (i = max-1; i >=0; i--)` seems to be correct.

**Comment by BackOrder** [comment]
What is [label_radius](https://github.com/Shoes3/shoes3/commit/589d234e6c7dc092eb2c7d0e5d24157724c779cc#diff-039163fedd17f4daf54936df4718d926R24) ? It seems to mean **label margin**, as in the margin around the label, which would be consistent with other Shoes styling names.

**Comment by ccoupe** [comment]
Read the [commit comment](https://github.com/Shoes3/shoes3/commit/589d234e6c7dc092eb2c7d0e5d24157724c779cc) . It is not a margin (measured in pixels). It's subtle and debatable which is better - a radius* multiplier or a radius+margin. I'm fairly certain that people that need this tweaking of position  can handle a multiplier.  Using a margin would on depend their font choice size (a future feature it we keep digging deeper and deeper) Inconvenient.  Drawing a larger chart size would require tweaking of label_radius to get them where wanted, Inconvenient.

Perhaps we can have both with a few extra days of work.. At the moment, I'd prefer to ship something soon, and find out if anyone cares enough to complain with practical usage.

**Comment by BackOrder** [comment]
I did read the commit comment but it only made sense after you explain it here. Always some concern about future backward compatibility and naming convention but **label_radius** could do for now. 

It should be noted that we should have a **label_margin** in the future that is actually independent of chart size as you mention larger chart means tweaking label_radius. It could be inconvenient if one simply maximize a window, Shoes code redrawing bigger but then also need to adjust label_radius.

> At the moment, I'd prefer to ship something soon, and find out if anyone cares enough to complain with practical usage.

Agreed.

**Comment by ccoupe** [comment]
Updated the [wiki](https://github.com/Shoes3/shoes3/wiki/Plot-Widget).  Getting everything into the manual will be challenge.

**Comment by BackOrder** [comment]
> Updated the wiki. 

The wiki page is very good. 

> Getting everything into the manual will be challenge.

Maybe someday #213 will see the day of light and we can use markdown directly, where wiki and manual will be totally compatible. 🍔

**Comment by ccoupe** [comment]
Considering, #293, it would be nice if we did have the e-book thing. We can get more a modern mingw for Windows so we can get better full_text gem,  if we give up RubyInstaller /devkit compatibility. RubyInstaller is a pretty low key project. 

We've got to face Ruby 2.3 and ruby 2.4 changes at the C level and the relentless march of updated dependent libraries. Probably Shoes 3.3.3.

**Comment by BackOrder** [comment]
> Considering, #293, it would be nice if we did have the e-book thing. 

I've been busy and no free time to work on this. It's actually not overly difficult. Though some concerns about the [Picky](https://github.com/floere/picky): it is working well but not actively maintained. One of the many problems with so many API these days, isn't it?

> We can get more a modern mingw for Windows so we can get better full_text gem, if we give up RubyInstaller /devkit compatibility. RubyInstaller is a pretty low key project.

My understanding is that RubyInstaller has alternatives as far as using Ruby goes but it is the only one to provide a fully working devkit. It would come to sense that if we would not rely on RubyInstaller Devkit, we come with a solution that can survive us, ensuring Shoes future (or at least easy for anyone to pick up where Shoes will be left off).

If you feel strongly against RubyInstaller, open an issue and let's talk about pros, cons, and potential solutions.

**Comment by ccoupe** [comment]
Actually, there is an earlier about Rubyinstaller, #281.

**Comment by dredknight** [comment]
Here is a fun custom chart that I did by using progress bar.

```Ruby
Shoes.app do

	def bar bar, dif
		anim=every(0.025) do |count|
			bar.fraction=count*(dif/(20-(15-0.75*count)))
			count == 20 ? anim.stop : nil
		end
	end
	
	flow width: 30, height: 200 do
		subtitle "Y", top: 90
	end
	for i in 0..9 do
		bar ( progress left: 30, top: 20+20*i, width: 150, height: 5 ), (rand(100).to_f/100)
	end
	stack top: 200, left: 0, width: 200 do
		subtitle "X", align: "center"
	end
end
```

**Comment by passenger94** [comment]
Yes !! Cool !!
No vertical progress bar that i'm aware of but you can probably create some funky ones from scratch with shapes for example

**Comment by BackOrder** [comment]
@dredknight very nice sample!

What do you need vertical progress bar for? If you feel strongly about it, open a new issue with a detailed description and tag it with enhancement. Otherwise you may give a try to @passenger94 's idea. Shouldn't be too hard - it's a box in a box. :)))

**Comment by dredknight** [comment]
Thanks @BackOrder and @passenger94 ! I was just messing with it nothing too special for the moment. 
I will definitely check the forms sometimes.

**Comment by ccoupe** [comment]
Closed in Shoes 3.3.2

**Comment by ccoupe** [comment]
Closed in Shoes 3.3.2
